File: backend/app/services/recruiter_service.py
"""
Recruiter service matching Java RecruiterService.java.
"""
from sqlalchemy.orm import Session

from app.exceptions.custom_exceptions import ResourceNotFoundException
from app.repositories.recruiter_repository import RecruiterRepository
from app.schemas.recruiter import RecruiterProfileResponse, RecruiterProfileUpdateRequest


class RecruiterService:
    """
    Recruiter-specific operations.
    Matches Java RecruiterService.java.
    """
    
    def __init__(self, db: Session):
        self.db = db
        self.recruiter_repo = RecruiterRepository(db)
    
    def get_my_profile(self, authenticated_user_id: int) -> RecruiterProfileResponse:
        """
        Get recruiter's own profile.
        Matches Java: @Transactional(readOnly = true) public RecruiterProfileResponse getMyProfile(...)
        """
        recruiter = self._find_by_user_id(authenticated_user_id)
        
        return RecruiterProfileResponse(
            user_id=recruiter.user.id,
            name=recruiter.user.name,
            email=recruiter.user.email,
            phone=recruiter.phone,
            designation=recruiter.designation,
            company_id=recruiter.company_id,
            company_name=recruiter.company.name if recruiter.company else None
        )
    
    def update_my_profile(
        self,
        authenticated_user_id: int,
        request: RecruiterProfileUpdateRequest
    ) -> RecruiterProfileResponse:
        """
        Update recruiter's own profile.
        Matches Java: @Transactional public RecruiterProfileResponse updateMyProfile(...)
        """
        recruiter = self._find_by_user_id(authenticated_user_id)
        
        # Update fields
        recruiter.phone = request.phone.strip() if request.phone and request.phone.strip() else None
        recruiter.designation = request.designation.strip() if request.designation and request.designation.strip() else None
        
        self.recruiter_repo.save(recruiter)
        self.db.commit()
        
        return self.get_my_profile(authenticated_user_id)
    
    def _find_by_user_id(self, user_id: int):
        """Find recruiter by user ID or raise exception."""
        recruiter = self.recruiter_repo.find_by_user_id(user_id)
        if not recruiter:
            raise ResourceNotFoundException("Recruiter profile not found.")
        return recruiter

File: backend/src/main/java/com/codeforge/service/RecruiterService.java
package com.codeforge.service;

import com.codeforge.dto.request.RecruiterProfileUpdateRequest;
import com.codeforge.dto.response.RecruiterProfileResponse;
import com.codeforge.entity.Recruiter;
import com.codeforge.entity.User;
import com.codeforge.exception.ResourceNotFoundException;
import com.codeforge.repository.RecruiterRepository;
import com.codeforge.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Service for recruiter profile management
 */
@Service
@RequiredArgsConstructor
public class RecruiterService {
    
    private final RecruiterRepository recruiterRepository;
    private final UserRepository userRepository;
    
    /**
     * Get recruiter profile for authenticated user
     */
    public RecruiterProfileResponse getMyProfile(Long userId) {
        Recruiter recruiter = recruiterRepository.findById(userId)
            .orElseThrow(() -> ResourceNotFoundException.notFound("Recruiter", "id", userId));
        
        return mapToRecruiterProfileResponse(recruiter);
    }
    
    /**
     * Update recruiter profile
     */
    @Transactional
    public RecruiterProfileResponse updateMyProfile(Long userId, RecruiterProfileUpdateRequest request) {
        Recruiter recruiter = recruiterRepository.findById(userId)
            .orElseThrow(() -> ResourceNotFoundException.notFound("Recruiter", "id", userId));
        
        User user = recruiter.getUser();
        
        if (request.getName() != null) {
            user.setName(request.getName());
        }
        
        if (request.getPhone() != null) {
            recruiter.setPhone(request.getPhone());
        }
        
        if (request.getPosition() != null) {
            recruiter.setPosition(request.getPosition());
        }
        
        if (request.getBio() != null) {
            recruiter.setBio(request.getBio());
        }
        
        userRepository.save(user);
        recruiter = recruiterRepository.save(recruiter);
        
        return mapToRecruiterProfileResponse(recruiter);
    }
    
    private RecruiterProfileResponse mapToRecruiterProfileResponse(Recruiter recruiter) {
        String companyName = recruiter.getCompany() != null ? recruiter.getCompany().getName() : null;
        Long companyId = recruiter.getCompany() != null ? recruiter.getCompany().getId() : null;
        
        return RecruiterProfileResponse.builder()
            .userId(recruiter.getId())
            .name(recruiter.getUser().getName())
            .email(recruiter.getUser().getEmail())
            .phone(recruiter.getPhone())
            .position(recruiter.getPosition())
            .bio(recruiter.getBio())
            .companyId(companyId)
            .companyName(companyName)
            .build();
    }
}
